sklearn/cluster/tests/test_k_means.py

```diff
@@ -892,4 +903,81 @@ def test_less_centers_than_unique_points():
     # centers have been used
     msg = ("Number of distinct clusters (3) found smaller than "
            "n_clusters (4). Possibly due to duplicate points in X.")
-    assert_warns_message(ConvergenceWarning, msg, k_means, X, n_clusters=4)
+    assert_warns_message(ConvergenceWarning, msg, k_means, X,
+                         sample_weights=None, n_clusters=4)
+
+
+def _sort_cluster_centers_and_labels(centers, labels):
+    sort_index = np.argsort(centers, axis=0)[:, 0]
+    sorted_labels = np.zeros_like(labels)
+    for i, l in enumerate(sort_index):
+        sorted_labels[labels == l] = i
+    return centers[sort_index, :], sorted_labels
+
+
+def test_weighted_vs_repeated():
+    # a sample weight of N should yield the same result as an N-fold
+    # repetition of the sample
+    sample_weights = np.random.randint(1, 5, size=n_samples)
+    X_repeat = np.repeat(X, sample_weights, axis=0)
+    for estimator in [KMeans(n_clusters=n_clusters, random_state=42),
+                      MiniBatchKMeans(n_clusters=n_clusters, batch_size=10,
+                                      random_state=42)]:
+        est_weighted = clone(estimator).fit(X, sample_weights=sample_weights)
+        est_repeated = clone(estimator).fit(X_repeat)
+        centers_1, labels_1 = _sort_cluster_centers_and_labels(
+                est_repeated.cluster_centers_, est_repeated.labels_)
+        centers_2, labels_2 = _sort_cluster_centers_and_labels(
+                est_weighted.cluster_centers_, np.repeat(est_weighted.labels_,
+                                                         sample_weights))
+        assert_almost_equal(v_measure_score(labels_1, labels_2), 1.0)
+        if not isinstance(estimator, MiniBatchKMeans):
```

Comment:
In case you are wondering why this test is not done for `MiniBatchKMeans`, that's because it fails. That is not due to the changes in this PR, however. I did a quick test on the master branch, comparing the `cluster_centers_` of `KMeans` and `MiniBatchKMeans`, and they are not the same.